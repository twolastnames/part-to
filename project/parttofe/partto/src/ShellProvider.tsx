import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "./App.scss";

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

export enum Theme {
  Default = "default",
  Dark = "dark",
}

export function changeTheme(theme: Theme) {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeChangeRef?.current && themeChangeRef.current(theme);
  themeNode.className = `${theme}Theme`;
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(Theme.Default);
  themeChangeRef = useRef(setTheme);
  return (
    <div id="theme" className="defaultTheme">
      <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
    </div>
  );
};
