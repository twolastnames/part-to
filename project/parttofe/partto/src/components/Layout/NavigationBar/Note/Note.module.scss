@import "../../../../App.scss";

.clickable {
  cursor: pointer;
}

.note {
  border: 1px solid var(--background-color);
  margin: var(--gap-medium);
  padding: var(--gap-medium);
  @media #{$portrait} {
    margin: var(--gap-small);
    padding: var(--gap-small);
  }
}

.heading {
  font: var(--font-strong-small);
}

.detail {
  font: var(--font-medium);
}
