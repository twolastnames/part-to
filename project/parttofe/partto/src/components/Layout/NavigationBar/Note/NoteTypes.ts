import { ReactNode } from "react";

export interface NoteProps {
  heading: ReactNode;
  detail: ReactNode;
  onClick?: () => void;
  sound?: string;
}
