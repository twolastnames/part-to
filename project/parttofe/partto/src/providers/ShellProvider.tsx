import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "../App.scss";
import {
  LeftDynamicItemSetPairProvider,
  RightDynamicItemSetPairProvider,
} from "./DynamicItemSetPair";
import { TimerProvider } from "./Timer";

export enum Theme {
  Default = "default",
  Dark = "dark",
}

const themeStorageKey = "theme";

export function getTheme(): Theme {
  return (localStorage.getItem(themeStorageKey) || Theme.Default) as Theme;
}

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

function setThemeClass() {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeNode.className = `${getTheme()}Theme`;
}

export function changeTheme(theme: Theme) {
  localStorage.setItem(themeStorageKey, theme.toString());
  themeChangeRef?.current && themeChangeRef.current(theme);
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(getTheme());
  themeChangeRef = useRef(setTheme);
  setTimeout(() => {
    setThemeClass();
  }, 0);
  return (
    <div id="theme" className="defaultTheme">
      <LeftDynamicItemSetPairProvider>
        <RightDynamicItemSetPairProvider>
          <TimerProvider>
            <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
          </TimerProvider>
        </RightDynamicItemSetPairProvider>
      </LeftDynamicItemSetPairProvider>
    </div>
  );
};
