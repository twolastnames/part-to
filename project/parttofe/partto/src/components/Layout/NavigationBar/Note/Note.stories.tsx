import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Note } from "./Note";
import { ShellProvider } from "../../../../ShellProvider";

const meta: Meta<typeof Note> = {
  component: Note,
  decorators: (Story) => (
    <ShellProvider>
      <Story />
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Note>;

export const Simple: Story = {
  args: {},
};
