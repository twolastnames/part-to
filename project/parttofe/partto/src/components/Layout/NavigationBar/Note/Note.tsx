import React from "react";

import classes from "./Note.module.scss";
import { NoteProps } from "./NoteTypes";
export function Note({ heading, detail, onClick }: NoteProps) {
  return (
    <div
      className={[classes.note, onClick ? classes.clickable : ""].join(" ")}
      onClick={() => onClick && onClick()}
      data-testid="Note"
    >
      <div className={classes.heading}>{heading}</div>
      <div className={classes.detail}>{detail}</div>
    </div>
  );
}
