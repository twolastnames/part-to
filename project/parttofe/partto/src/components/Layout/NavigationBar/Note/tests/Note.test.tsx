import React from "react";
import { expect, test } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { ShellProvider } from "../../../../../ShellProvider";
import { Note } from "../Note";

test("snapshot", () => {
  render(
    <ShellProvider>
      <Note />
    </ShellProvider>,
  );
  const component = screen.getByTestId("Note");
  expect(component).toMatchSnapshot();
});
