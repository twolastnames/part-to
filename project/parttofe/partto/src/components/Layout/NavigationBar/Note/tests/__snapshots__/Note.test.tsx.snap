// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="note "
  data-testid="Note"
>
  <div
    class="heading"
  >
    Due North
  </div>
  <div
    class="detail"
  >
    $500 for fancy cars, $20 if it looks like you're poor
  </div>
</div>
`;
